### ISSUE TITLE
Proposal for future code reorganization

### ISSUE THREAD

**Issue by hgrecco** [hgrecco/pint#1466]
I would like to propose a code reorganization after the next release and before we start the next cycle. The goal is to group things that interact strongly in order to make reasoning about the code easier. Hopefully this will lead to isolate better the different components avoiding side effects, and making it easier to contribute new code.

I am not sure that this is a good idea, I would appreciate some feedback.

Create folders by Registry feature
- base:
  - registry.py: BaseRegistry
  - parsing.py: PrefixDefinition, UnitDefinition, DimensionDefinition, AliasDefinition, DefaultsDefinition
  - loading.py: functions that takes a BaseRegistry and any of the classes defined in parsing
- context
  - registry.py: ContextRegistry
  - parsing.py: ContextDefinition
  - loading.py: Context, functions that takes a SystemRegistry and any of the classes defined in parsing
- systems:
  - registry.py: SystemRegistry
  - parsing.py: GroupDefinition, SystemDefinition
  - loading.py: Group, System, functions that takes a SystemRegistry and any of the classes defined 
in parsing

In registry.py leave UnitRegistry, LazyRegistry, ApplicationRegistry

Not clear:
1. NonMultiplicativeRegistry: own folder or in base? Their definitions are intertwined now with the base Definitions. Should we have them all together like systems and groups or subclass the UnitDefinition?
2. Converters: depending on what we do in (1)
3. Systems and Groups: should we split them into SystemRegistry and GroupRegistry
4. Move measurement to this type of organization (i.e. MeasurementRegistry)
5. Move numpy support to this type of organization (i.e. NumpyRegistry)

**Comment by hgrecco** [comment]
I prototyped this and looks quite nice. Smaller files, smaller more focused classes, better understanding about the connections between them. But it is a very large reorganization 

After #1464 is merged, maybe I will push it to a separate branch so you can give me your opinion.

ping @jules-ch @keewis @dalito @andrewgsavage

**Comment by andrewgsavage** [comment]
I agree, it is difficult for a newcomer to get their head around the registry code. 

This looks like a good idea, will be good to see it.

**Comment by dalito** [comment]
I also think restructuring the code is a good idea. After a longer break it took quite a while to understand the code again.

**Comment by hgrecco** [comment]
Just FYI, I prototyped the separation of three "facets" (Group, System, Context). I think it looks quite good. What do you think? I will do others soon.

**Comment by hgrecco** [comment]
There are now five facets:
- base (this is half-baked now, and I think it will be quite large)
- context
- group
- nonmultiplicative
- system
 
Some nice things:
- Files are smaller
- Strong coupling within a facet
- Little coupling between facets with few imports between them, which simplifies understanding the code. 

and still some more decoupling could be done.

Why this is an initial attempt, is large enough to trigger a more informed discussion.
Thinks we could do:
- Split the Quantity class and organized them in a subclassing hierarchy, just as we do for the Registry
- Split measurement (i.e. MeasurementRegistry)
- Split numpy support  (i.e. NumpyRegistry)
- Split string parsing and printing.

**Comment by hgrecco** [comment]
Now that 0.19 is out, I will proceed to reorganize the code. If you have any suggestion, please let me know. You can check out a sketch of the current proposal in https://github.com/hgrecco/pint/pull/1479

**Comment by hgrecco** [comment]
I have just pushed a concrete code reorganization proposal.

I have divided the code into 9 facets: plain, nonmultiplicative, measurement, formatting, numpy, dask, group, system, context. Each facet encapsulate a feature (or tight connection of features). It achieves this by providing a registry like class that can serve a base class for a more complex and feature rich registry. Some facets also implement a Quantity and Unit class, which are also joined seamlessly to create the Quantity class.

I think the outcome is quite nice. The code is divided into smaller pieces, becoming more hackable. The import section in each of the modules becomes more focused and the connection between modules more obvious and transparent. Also, the helper functions can be now put together with the actual code using them in the fact.

There is still work to do. This refactoring showed some part of the code which is not in use anymore as well as some inconsistency in the typing. I also want to reduce the exports from each facet. But I would really appreciate your feedback about the direction, please check #1479 

ping @jules-ch @keewis @dalito @andrewgsavage

**Comment by hgrecco** [comment]
I will merge this by the end of the week. There are still some pending things to do:
1. move certain code into facets (e.g. babel, formatting)
2. move `__init__` arguments to the right facet registry (e.g. `fmt_locale` should be in the `FormattingRegistry`)
3. check again if some dead code could be removed.
4. Fix some bugs introduced in the docs due to this change.

And a few decisions to make:
1. shall we include explicitly the bases that are needed. (e.g. `NonmultiplicativeRegistry` requires `PlainRegistry`, do we make it inherit from it or leave it to the user?)
2. Is there a way to handle `__init__` arguments better? Dataclasses? attr? other?

**Comment by hgrecco** [comment]
This is ready for merging. Still some pending improvements, but I think is worth merging now so next commits build up on this.

**Comment by andrewgsavage** [comment]
> 1. shall we include explicitly the bases that are needed. (e.g. `NonmultiplicativeRegistry` requires `PlainRegistry`, do we make it inherit from it or leave it to the user?)

It isn't obvious from reading pint/facets/nonmultiplicative/registry.py `NonmultiplicativeRegistry` requires `PlainRegistry` so I agree something more is needed. Inheriting would make this clearer, or a note in the docstring if you want to avoid linking between facets.

**Comment by hgrecco** [comment]
@andrewgsavage I agree. I will start by inheriting, as I do not see a problem with linking facets (I am trying to follow some import rules that I will post in another issue and maybe build a linter, but later).

In the future, we can discuss if non-inheriting is a good idea. The only reason I can see now is that somebody can build a different PlainRegistry and try it out. But I think this will not happen.

**Comment by hgrecco** [comment]
I have decided to merge it in its current form, and I will add the inheritance in the next commit
